src/common/base/CuckooHashCache.h

```diff
@@ -0,0 +1,875 @@
+    // checks that the key isn't already in the table. cuckoo hashing presents
+    // multiple concurrency issues, which are explained in the function. The
+    // following return states are possible:
+    //
+    // ok -- Found an empty slot, locks will be held on both buckets after the
+    // function ends, and the position of the empty slot is returned
+    //
+    // failure_key_duplicated -- Found a duplicate key, locks will be held, and
+    // the position of the duplicate key will be returned
+    //
+    // failure_table_full -- Failed to find an empty slot for the table. Locks
+    // are released. No meaningful position is returned.
+    template <typename K>
+    table_position cuckoo_insert(const hash_value hv, TwoBuckets &b, K &key) {
+        int res1, res2;
+        if (!try_find_insert_bucket(b.i1, res1, hv.partial, key)) {
+            return table_position{b.i1, static_cast<size_t>(res1),
+                                  failure_key_duplicated};
+        }
+        if (!try_find_insert_bucket(b.i2, res2, hv.partial, key)) {
+            return table_position{b.i2, static_cast<size_t>(res2),
+                                  failure_key_duplicated};
+        }
+        if (res1 != -1) {
+            return table_position{b.i1, static_cast<size_t>(res1), ok};
+        }
+        if (res2 != -1) {
+            return table_position{b.i2, static_cast<size_t>(res2), ok};
+        }
+        return table_position{0, 0, failure_table_full};
+    }
+
+    // add_to_bucket will insert the given key-value pair into the slot. The key
+    // and value will be move-constructed into the table, so they are not valid
+    // for use afterwards.
+    template <typename K, typename... Args>
+    void add_to_bucket(const size_t bucket_ind, const size_t slot,
+                       const partial_t partial, K &&key, Args &&... val) {
+        uint32_t e = nebula::time::WallClock::fastNowInSec() + ttl_;
+        buckets_.setKV(bucket_ind, slot, partial, e, std::forward<K>(key),
+                       std::forward<Args>(val)...);
+        ++locks_[lock_ind(bucket_ind)].elem_counter();
+    }
+
+    // try_find_insert_bucket will search the bucket for the given key, and for
+    // an empty slot. If the key is found, we store the slot of the key in
+    // `slot` and return false. If we find an empty slot, we store its position
+    // in `slot` and return true. If no duplicate key is found and no empty slot
+    // is found, we store -1 in `slot` and return true.
+    template <typename K>
+    bool try_find_insert_bucket(const size_t bucket_ind, int &slot,
+                                const partial_t partial, const K &key) {
+        bucket &b = buckets_[bucket_ind];
+        slot = -1;
+        for (int i = 0; i < static_cast<int>(slot_per_bucket()); ++i) {
+            if (try_reclaim(bucket_ind, i)) {
+                slot = i;
+            } else if (b.occupied(i)) {
+                if (!is_simple() && partial != b.partial(i)) {
```

Comment:
We could eliminate the `if` here. Because we could ensure `is_simple` at compile time.